xlators/cluster/dht/src/dht-rename.c

```
@@ -1385,12 +1386,36 @@ dht_rename_create_links(call_frame_t *frame)
      * unlink the newname we created, we would have effectively lost the
      * file to rename operations. */
     if (dst_hashed != src_hashed && src_cached != dst_hashed) {
-        gf_msg_trace(this->name, 0, "linkfile %s @ %s => %s", local->loc.path,
-                     dst_hashed->name, src_cached->name);
+        dict_t *xattr_new = NULL;
+
+        xattr_new = dict_copy_with_ref(xattr, NULL);
+        if (!xattr_new) {
+            ret = -1;
+            goto cleanup;
+        }
+
+        /* rename(src,dst): if a stale linkto to src is present in the dst
+         * hashed, linkto creation fails with EEXIST.
+         * handling: populate the MKNOD REPLACE key for "forcing" posix the
+         * creation of the linkto with correct gfid. */
+        ret = dict_set_str(xattr_new, GF_MKNOD_REPLACE_KEY, "yes");
+        if (ret < 0) {
+            gf_msg(this->name, GF_LOG_WARNING, 0, DHT_MSG_DICT_SET_FAILED,
+                   "Failed to set dictionary value: key = %s,"
+                   " path = %s",
+                   GF_MKNOD_REPLACE_KEY, local->loc.path);
+            ret = 0;
```

Are you sure we can ignore this error ? in most of the cases the linkto file won't exist, so the presence of this key is irrelevant. But when the linkto file exists, failing to set this key will cause failure of the rename. I think it's better to fail here with a more consistent error, like ENOMEM, than a later failure with ESTALE or EEXIST which is harder to understand by the user.

We also fail if `xattr_new` cannot be allocated, so I think we should also fail if the key cannot be added.

---

Actually, I had doubts about it, as it handles rare cases i was not sure if we should fail the fop.
But i agree that ENOMEM is a more consistent error and a failure in the infrastructure should be rare too..